kmath-geometry/src/commonMain/kotlin/scientifik/kmath/geometry/EuclideanSpace.kt

```diff
@@ -0,0 +1,57 @@
+package scientifik.kmath.geometry
+
+import scientifik.kmath.dimensions.*
+import scientifik.kmath.operations.ExtendedField
+import scientifik.kmath.operations.Field
+import scientifik.kmath.operations.invoke
+import scientifik.kmath.structures.asSequence
+
+internal data class VectorImpl<T, D : Dimension>(private val array: List<T>) : Vector<T, D> {
+    override val size: Int get() = array.size
+
+    override fun get(index: Int): T = when {
+        index < 0 || index >= size -> error("Accessing outside of point bounds")
+        else -> array[index]
+    }
+
+    override fun iterator(): Iterator<T> = array.iterator()
+}
+
+fun <T> vectorOf(a1: T): Vector<T, D1> = VectorImpl(listOf(a1))
+fun <T> vectorOf(a1: T, a2: T): Vector<T, D2> = VectorImpl(listOf(a1, a2))
+fun <T> vectorOf(a1: T, a2: T, a3: T): Vector<T, D3> = VectorImpl(listOf(a1, a2, a3))
+fun <T> vectorOf(a1: T, a2: T, a3: T, a4: T): Vector<T, D4> = VectorImpl(listOf(a1, a2, a3, a4))
+fun <T> vectorOf(a1: T, a2: T, a3: T, a4: T, a5: T): Vector<T, D5> = VectorImpl(listOf(a1, a2, a3, a4, a5))
+
+private fun <T> Field<T>.square(a: T): T = a * a
+
+class EuclideanSpace<T : Any, D : Dimension>(
+        override val dim: D,
+        override val field: ExtendedField<T>
+) : InnerProductSpace<T, D, Vector<T, D>> {
+    private val dimSize = dim.dim.toInt()
+
+    override fun distance(a: Vector<T, D>, b: Vector<T, D>): T =
+            field.sqrt(a.asSequence().zip(b.asSequence()).map { (l, r) -> field { square(l - r) } }.reduce(field::add))
+
+    override fun norm(a: Vector<T, D>): T =
+            field.sqrt(dotProduct(a, a))
+
+    override fun dotProduct(a: Vector<T, D>, b: Vector<T, D>): T =
+            a.asSequence().zip(b.asSequence()).map { (l, r) -> field.multiply(l, r) }.reduce(field::add)
```

Comment:
All sequence operation are performance pitfalls. Need to review or benchmark that.

Reply:
yeah, in other places I have seen the same achieved with something like `fun produce(supplier: (Int) -> T): Vector<T>`. Think I will replace here (at least for the sake of API consistency). Also `produce`-approach is better, since size of the vector would be controlled by `VectorSpace` not `Sequence<T>`.